File: api/src/main/java/fr/huiitre/tools/modules/core/module/application/usecase/UpdateModuleUseCase.java
package fr.huiitre.tools.modules.core.module.application.usecase;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.core.module.application.command.UpdateModuleCommand;
import fr.huiitre.tools.modules.core.module.application.ports.ModuleRepository;
import fr.huiitre.tools.modules.core.module.domain.Module;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;

@Service
@Transactional
public class UpdateModuleUseCase implements SecuredUseCase {

    private final static Logger logger = LoggerFactory.getLogger(UpdateModuleUseCase.class);

    @Override
    public RoleCode requiredRole() {
        return RoleCode.ADMIN;
    }

    private final ModuleRepository moduleRepository;

    public UpdateModuleUseCase(
            ModuleRepository moduleRepository) {
        this.moduleRepository = moduleRepository;
    }

    public void execute(Long moduleId, UpdateModuleCommand command) {

        // * vérification si le module existe */
        Module module = moduleRepository.findById(moduleId)
                .orElseThrow(() -> new IllegalArgumentException("MODULE_NOT_FOUND"));

        // * mise à jour du module */
        module.update(
                command.getCode(),
                command.getName(),
                command.getDescription(),
                command.getActive());

        moduleRepository.update(module);
    }
}

File: api/Modules/Core/Access/Application/Usecases/UpdateModuleUseCase.cs
using Tools.Api.Modules.Core.Access.Application.Ports;
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Realtime.Application;
using Tools.Api.Modules.Core.Realtime.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Ports;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;

namespace Tools.Api.Modules.Core.Access.Application.Usecases;

// Cas d'usage administrateur : modifier un module fonctionnel, activation comprise.
public sealed class UpdateModuleUseCase(
    UseCaseAuthorizer authorizer,
    IModuleRepository moduleRepository,
    ILogger<UpdateModuleUseCase> logger,
    RealtimeEventService realtimeEventService
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.Admin;

    public async Task Execute(UpdateModuleCommand command)
    {
        if (!await moduleRepository.ExistsAsync(command.ModuleId))
        {
            throw AppException.NotFound("MODULE_NOT_FOUND", "Module introuvable.");
        }

        var code = command.Code.Trim();

        // Le module qu'on modifie est exclu du contrôle : garder son propre code n'est pas
        // un conflit.
        if (await moduleRepository.CodeExistsAsync(code, command.ModuleId))
        {
            throw AppException.Conflict("MODULE_CODE_ALREADY_EXISTS", "Ce code de module est déjà utilisé.");
        }

        await moduleRepository.UpdateAsync(
            command.ModuleId,
            code,
            command.Name.Trim(),
            command.Description?.Trim(),
            command.Active);

        // L'activation est la modification qui a un effet visible pour les utilisateurs :
        // elle mérite d'apparaître explicitement dans la trace.
        logger.LogInformation(
            "Module modifié par userId={ActorId} : moduleId={ModuleId} code={Code} actif={Active}",
            CurrentUser.UserId,
            command.ModuleId,
            code,
            command.Active
        );

        try
        {
            // Toucher le module lui-même concerne tous ses membres, pas un seul utilisateur —
            // TargetModuleId résout la liste, on ne la connaît pas ici.
            await realtimeEventService.PublishAsync(
                PublishRealtimeEventCommand.ForModule(command.ModuleId, "Core.ModuleUpdated"));
        } catch(Exception ex)
        {
            logger.LogWarning(ex, "Push temps réel de la mise à jour du module échoué pour moduleId={ModuleId}", command.ModuleId);
        }
    }
}
